boot: improve memory region sanity checks

axel-h opened the pull request:
- Adjust free memory to kernel window, as well for the lower as the upper limit.
- Log more detailed boot message to help porting.

axel-h committed:
boot: improve memory region sanity checks

- Adjust free memory to kernel window, as well for the lower as the
  upper limit.
- Log more detailed boot message to help porting.

Co-authored-by: Gerwin Klein <gerwin.klein@proofcraft.systems>
Signed-off-by: Axel Heider <axelheider@gmx.de>
Signed-off-by: Gerwin Klein <gerwin.klein@proofcraft.systems>

kent-mcleod commented:
Are you able to explain more why you think the previous code was problematic?
`check_available_memory` checks if the declared memory regions are valid and if not returns an error that aborts the boot.  The new warnings diagnostics if memory regions are out of the kernel window will create many spurious warnings on 32-bit architectures that have lots of memory outside of the kernel window which shouldn't be a warning.  If you have different contexts where this should be a warning, then you could add warnings in the root tasks when it processes the initial untyped list.

axel-h commented:
This produces output that is helpful for porting - and thus falls into the category that we lack a mechanism to have more verbose output for this case and allow turning this off otherwise. I agree that on 32-bit this might be annoying due to the address space limitation.
Aside from this:
- `PADDR_BASE` is checked also, not just `PADDR_TOP`. Matters only when `PADDR_BASE` is not 0.
- catch the case where `n_available and `avail_reg[MAX_NUM_FREEMEM_REG]` size differs. This usually indicates a porting or configuration problem.
- things are consolidated in `check_available_memory()`.

axel-h commented:
Don't prefix the message with "WARNING" and add comments in the code that this may happen on 32-bit architectures.

lsf37 commented on src/kernel/boot.c:
```diff
@@ -839,24 +834,63 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+         * small to support a kernel windows that is big enough.
+         */
+        if ((r->start > PADDR_TOP) || (r->end <= PADDR_BASE)) {
+            printf("    can't use region, outside of kernel window\n");
```
If I understand the boot code correctly, removing things from `freemem` will mean that they are used for device untyped. So the memory is still used, just not available for other kernel objects. 

```suggestion
            printf("    Region outside of kernel window, will use for device untypeds instead.\n");
```

lsf37 commented on src/kernel/boot.c:
```diff
@@ -810,26 +810,21 @@ BOOT_CODE void bi_finalise(void)
     }
 
+    word_t cnt = 0;
     printf("available phys memory regions: %"SEL4_PRIu_word"\n", n_available);
```
Not from this PR, but if we're changing the logging, we might keep the capitalisation consistent with the `Bootstrapping kernel` etc messages, which start with upper case.

```suggestion
    printf("Available phys memory regions: %"SEL4_PRIu_word"\n", n_available);
```

lsf37 commented on src/kernel/boot.c:
```diff
@@ -839,24 +834,63 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+            /* Region is partial addressable only, this is also not uncommon on
+             * 32-bit architectures due to the limited virtual address space.
+             */
+            printf("    kernel window addressable:"
```
```suggestion
            printf("    Kernel window addressable:"
```

lsf37 replied:
(happy to leave that one as is, see below)

lsf37 reviewed (dismissed):
This looks Ok from my side, I only have a few wording tweaks to indicate that the memory is not fully lost, it's just used for device untypeds instead (+ making capitalisation consistent).

I was about to suggest to print `)` instead of `]` for the regions (see #1008), but it's probably better to do that consistently in a separate PR.

We should do a hardware run and check the logs to see how verbose the new output is. Will kick that off now.

lsf37 commented:
Doesn't seem to be bad at all, the only one I randomly saw in the logs was for AM335X (beagle bone black):

```
Bootstrapping kernel
  available phys memory regions: 1
    [80000000..a0000000]
      kernel window addressable: [80000000..9ff00000]
```

Seeing it like that, the lower-level `kernel` in the log message does make sense, so I'd be fine with just leaving it as is.

lsf37 commented on src/kernel/boot.c:
```diff
@@ -839,31 +834,80 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+            return 0;
+        }
+
+        /* Sanitize the available memory region, only such memory can be used
+         * for kernel objects that is visible in the kernel window. On 32-bit
+         * architectures loosing some memory here is not uncommon, because the
+         * 32-bit virtual address space is too small to support a kernel window
+         * that is big enough. Memory outside of he kernel window is not list,
+         * it is made available as device untyped memory.
+         */
```
This should be rephrased a bit. Maybe:

> Sanitize the available memory region: only memory in the kernel window can be used for kernel objects. On 32-bit architectures it is not uncommon for provided regions for provided regions to be only partially in the kernel window, because the available address space is small. Memory outside of the kernel window is made available as device untyped memory.

axel-h replied:
done

lsf37 reviewed (commented):
The protocol is usually to not make any more substantial changes after approval, but just to implement review feedback. New changes should go into a new PR.

In this case, I'm fine with the change, but have some tweaks for comments.

axel-h commented on src/kernel/boot.c:
```diff
@@ -839,31 +834,80 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
 
         /* Regions must be ordered and must not overlap. Regions are [start..end),
-           so the == case is fine. Directly adjacent regions are allowed. */
+         * so the '=='-case is fine. Directly adjacent regions are allowed.
```
done

lsf37 commented on src/kernel/boot.c:
```diff
@@ -896,31 +891,79 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+            printf("    [%"SEL4_PRIx_word"..%"SEL4_PRIx_word"] kernel window addressable\n",
+                   usable_reg.start, usable_reg.end);
+            if (usable_reg.end != r->end) {
+                printf("    [%"SEL4_PRIx_word"..%"SEL4_PRIx_word"] device untyped memory\n",
```
We have since changed the printing to be explicit about the end of the region being exclusive, i.e. `[..)`. I'll fix that up.

lsf37 commented on src/kernel/boot.c:
```diff
@@ -896,31 +891,79 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+         */
+        if ((r->start > PADDR_TOP) || (r->end <= PADDR_BASE)) {
+            printf("    region outside of kernel window, available as device untypeds.\n");
+            continue;
```
`continue` is fine for the C parser, but is a problem for the binary verification tool chain. We might need to to slightly uglify this to avoid `continue`.

lsf37 commented on src/kernel/boot.c:
```diff
@@ -896,31 +891,79 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+             */
+            printf("    WARNING: can't use region, avail_reg[] is full\n");
+            assert(0);
+            continue;
```
(same `continue` problem here)

lsf37 commented on src/kernel/boot.c:
```diff
@@ -896,31 +891,79 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+             */
+            if (usable_reg.start != r->start) {
+                printf("    [%"SEL4_PRIx_word"..%"SEL4_PRIx_word"] device untyped memory\n",
+                       r->start, usable_reg.end);
```
Small bug:

```suggestion
                printf("    [%"SEL4_PRIx_word"..%"SEL4_PRIx_word"] device untyped memory\n",
                       r->start, usable_reg.start);
```

lsf37 replied:
We should also call it "potential device untyped memory", because if it overlaps with a reserved region it will not become device untyped and the message will be confusing when it then doesn't appear as device untyped in the end.

lsf37 commented on src/kernel/boot.c:
```diff
@@ -896,31 +891,79 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+         * window, because the available address space is small. Memory outside
+         * of the kernel window is made available as device untyped memory.
+         */
+        if ((r->start > PADDR_TOP) || (r->end <= PADDR_BASE)) {
```
This can't be right -- a region `[PADDR_TOP, PADDR_TOP+0x10)` would become a `usable_reg = [PADDR_TOP, PADDR_TOP)` below, which is empty. (Hey, did I ever mention how cool inclusive-end regions are, they can't be empty..).

I think we want this:

```suggestion
        if ((r->start >= PADDR_TOP) || (r->end <= PADDR_BASE)) {
```

Still need to double-check if the end part is correct, but I think it is.

lsf37 replied:
(I think `add_mem_p_regs` has the same bug, but it is harmless to produce empty regions there. Here it will produce confusing output)

lsf37 commented on src/kernel/boot.c:
```diff
@@ -867,27 +867,22 @@ BOOT_CODE void bi_finalise(void)
 }
 
-BOOT_CODE static inline pptr_t ceiling_kernel_window(pptr_t p)
+BOOT_CODE static word_t check_available_memory(word_t n_available,
```
The name `check_available_memory` is no longer right. It no longer just checks, it populates `avail_reg`. I would propose `init_avail_reg`.

lsf37 replied:
The more I think about it, the less I like that this function is no longer pure. I might have a go at splitting it into two parts (pure check and then init) and see if that looks any better. It would mean going through the regions twice, but these are not large arrays, so that should not be a problem, and thinking about verifying this in the future, having all of these checks separate from the state effects is probably nicer.

midnightveil commented on src/kernel/boot.c:
```diff
@@ -867,27 +867,22 @@ BOOT_CODE void bi_finalise(void)
-     */
-    if (pptr_to_paddr((void *)p) > PADDR_TOP) {
-        p = PPTR_TOP;
+    /* Ensure avail_reg is properly initialized. */
```
Is this actually necessary if `REG_EMPTY` is all 0?

lsf37 replied:
It's not and it would be semantically fine to leave out. I was even about to remove it, but then I remembered that I don't want to have to reason in verification that it's still all 0 when this function gets called, so I left it in.

lsf37 replied:
Now that you made me think about it more, we can actually leave it out: we never need to know what the previous values of `avail_reg` are, we only write to it and we only care about the indices that we have written to. Will remove.

lsf37 reviewed (approved):
I would now be happy with the outcome (after commits are squashed), but given that I tweaked a lot of this, it'd be good to get a second review.

lsf37 committed:
x86,boot_sys.c: fix off-by-one in region check

Fix a (harmeless) off-by-one error in the kernel window check.
PADDR_TOP is an exclusive end value, so if reg.start == PADDR_TOP, then
reg.start is already outside the kernel window.

Previously, this case would have created an empty region that would
later have been ignored. It would have caused a spurious region print
and consumed one slot in the region array, but that is all.

Signed-off-by: Gerwin Klein <gerwin.klein@proofcraft.systems>

Indanz commented on src/kernel/boot.c:
```diff
@@ -889,40 +878,92 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
     return true;
 }
 
+/* We can't declare arrays on the stack, so this is space for init_avail_reg()
+ * below to populate and init_freemem() to consume. */
+BOOT_BSS static region_t avail_reg[MAX_NUM_FREEMEM_REG];
```
Normally I would pass `avail_reg` as an argument to `init_avail_reg()` instead of sharing a global variable like this. I had to look twice to see what was going on, it's a bit unexpected.

lsf37 replied:
Yes, there's not much we can do about this one. Arrays on the stack don't work on the verification side, because they create pointers into the stack (which the memory model disallows). Passing it as an argument just shifts it to the caller stack. I guess we could pass the global as an argument, but that also is a bit strange.

lsf37 replied:
(This is old, though, the commit just moves it up to the new function)

Indanz replied:
> I guess we could pass the global as an argument, but that also is a bit strange.

Better than what we have now. Then at least it's conceptually on the stack of `init_freemem()` and the comment makes sense. Now you've split of `init_avail_reg()`, it's a true global variable and can't be on the stack of either function.

(If you put the array inside a struct you probably could put it on the stack, but it might be too big for that anyway.)

midnightveil replied:
I do somewhat agree - several of these boot functions depend on global state (e.g. the ni_sdk_boot) and it could pass data around explicitly. But it's really not that bad, only when it goes cross-file.

Struct-with-array-on-stack has same problem as array-on-stack: kernel C subset does not permit taking refernces to locations on the stack.

Unless the proposal is to pass the struct-with-array by value, so that you don't have to take the address of it. But ew, that's a lot of data copying.

lsf37 replied:
Just tried this out and it's not great. The ARRAY_SIZE macro doesn't work on an array that is a parameter, because that just decays to a pointer. This is the same problem we'll have in verification when we get here, because we'd need to pass through precondtions about the size of this array and the fact that it actually is an array etc, which just complicates things over a single-use global array. Usually globals would be bad, because they can be modified from anywhere, but in this kind of scenario they are easy because we never read from this global again after init_free_mem.

Edit: this overlapped. "this" = my proposal to pass the global as a parameter.

lsf37 replied:
I agree that wrapping in a struct is not going to work.

midnightveil replied:
Yeah, makes sense. It's fine.

Indanz commented on src/kernel/boot.c:
```diff
@@ -889,40 +878,92 @@ BOOT_CODE static bool_t check_available_memory(word_t n_available,
+            }
+        }
+    }
+
```
Stray empty new line.

Indanz reviewed (approved):
Just some nitpicks, otherwise looks fine.
